### DISCUSSION TITLE
Update and Downgrade Liquidsoap

### DISCUSSION THREAD

**Discussion by Marcioghiraldelli** [AzuraCast/AzuraCast#6668] [Ideas]
Today in my spare time, I was studying Azuracast more and more so that I could start contributing to the tool in the near future.
And I have a doubt, if I have an older version of Azuracast and I want to update only liquidsoap, I can't because Azuracastr forces me to update everything.
So I did some tests, I even managed, for example, in an older version of Azuracast to update to the latest stable version of liquidsoap, but the stations don't start because the current version of LS has strings and variables that my Azuracast version doesn't apply.
So I thought who should change the src/Radio/Backend/Liquidsoap/ConfigWriter.php file where it seems to create the liquidsoap files for each station.
I haven't done this part yet, I even thought about using the current version of github because as in the test only the azuracast web base is older plus ls and src/Radio/Backend/Liquidsoap/ConfigWriter.php no, it could be that right or am I wrong?

note; I know that the right thing to do is to mount an image of ls with the changes because if you restart the container you will return to the previous ls.

**Comment by Moonbase59** [comment]
Taking into account the rapid development (and breaking changes!) of Liquidsoap, doing this in a well-documented and supported way is extremely difficult if not impossible. Just too much might break a lot of stations that should run!

For testing, you can always open a bash in the container and try out other LS versions in there (or adapt ConfigWriter). I’ve been doing that lots of times, to test out stuff and make suggestions. Then again, as you say, it’ll all be gone with the next container restart…

**Reply by Marcioghiraldelli** [reply]
Yes, that I know.
But I believe that if you make a personalized image and point it inside the container, when you restart it, it won't disappear, but my doubt is not that, but rather the part of adapting a new LS to an old version.
I can install the new LS easily, but the stations stop working, so I thought about getting the current ConfigWriter file, from the latest version, and replacing the old ConfigWriter. It might work, even as a test, right? Well, I haven't done that yet.
It seems that the only problem I face are the variables and strings that change in the latest version of LS.

**Reply by Moonbase59** [reply]
Yeah, but… not for the faint-of-heart ;-) You might break lots of stuff.

Then again: Yes, I have made changes to ConfigWriter in my installation when we were hunting bugs, and you actually _can_ do it.

**A hint** for quickly working up patches _within_ the container: Have a mini editor like `nano` installed (per default, there is none).

In `/var/azuracast/azuracast.env`, at the end of the file, include it like so:
```
# Extra Ubuntu packages to install upon startup
# Standard: "Separate package names with a space. Packages will be installed during container startup."
INSTALL_PACKAGES_ON_STARTUP=nano
```

which will actually fetch and install `nano` from the repos on every container start (and produce some error messages, since not all locales are installed), so at least you have something to do a quick edit. ;-)

**Hint 2:** This is also the place where you can install debuggers like `gdb`. If debugging a new Liquidsoap, @toots from savonet might ask you for debugger output.

**Hint 3:** After `wget`ting or `curl`ing a new Liquidsoap, I prefer installing the `.deb`s using `apt install ./liquidsoap-xxx.deb` (need to be root within the container), because `apt` is clever enough to remove the old and install the new version in one go. (You _can_ of course also use `dpkg` and do it manually.)  Just ignore any errors about missing man pages, and _always_ check if the new Liquidsoap version is running, using `liquidsoap --version`.

Then just go to your AzuraCast web UI and on the profile page, do a STOP, followed by a START on the "AutoDJ Service" card. Always _wait_ until the web UI confirms it’s actually stopped, or started.

For copying files around between linux machines (your host, your workstation, …), you can use tools like `scp` anytime from within the container.
